File: tools/remove_empty_authors.py
# This scripts deletes all authors that have no books associated with them.
import requests

# Configuration constants
ABS_HOST = ""  # AudiobookShelf Host URL
LIBRARY_ID = None  # Leave None to process all libraries you have
API_KEY = ""  # API Key from user settings
DELETE_WITHOUT_CONFIRMATION = False  # Set to True to delete authors without confirmation


library_ids = []
if LIBRARY_ID is None:
    all_libraries_response = requests.get(f"{ABS_HOST}/api/libraries?token={API_KEY}")
    for library in all_libraries_response.json()['libraries']:
        library_ids.append(library['id'])
else:
    library_ids.append(LIBRARY_ID)

print(f"Processing {len(library_ids)} libraries")

for library_id in library_ids:
    print(f"Processing library {library_id}")
    authors_response = requests.get(f"{ABS_HOST}/api/libraries/{library_id}/authors?token={API_KEY}")
    authors_to_delete = []
    for author in authors_response.json()['authors']:
        author_id = author['id']
        author_name = author['name']
        author_num_books = author['numBooks']
        if author_num_books == 0:
            authors_to_delete.append((author_id, author_name, author_num_books))
    print(f"\n\nFound {len(authors_to_delete)} authors without books")
    print("The following authors will be deleted:")
    for author in authors_to_delete:
        print(f"{author[1]} ({author[2]} books) - ID: {author[0]}")
    if not DELETE_WITHOUT_CONFIRMATION:
        user_input = input("Do you want to delete these authors? (y/n): ")
        if user_input.lower() != "y":
            print("Skipping deletion")
            continue
    print("\n\n----------------------\n\n")
    for author in authors_to_delete:
        author_id = author[0]
        delete_response = requests.delete(f"{ABS_HOST}/api/authors/{author_id}?token={API_KEY}")
        if delete_response.status_code == 200:
            print(f"Deleted author {author[1]}")
        else:
            print(f"Error deleting author {author[1]}: {delete_response.status_code}")
    print("\n\n----------------------\n\n")

print("Done")

File: web/src/tools/removeEmptyAuthors.ts
import { fetchLibraryIds } from '@/common/libraryIds'
import { useApi } from '@/composables/useApi'
import type { ToolResult } from '@/types/tool'

async function getAllLibraries() {
  const { get } = useApi()
  try {
    const response = await get('/api/libraries')
    return response.data.libraries || []
  } catch (error) {
    console.error('Error fetching libraries:', error)
    return []
  }
}

async function getLibraryAuthors(libraryId: string) {
  const { get } = useApi()
  try {
    const response = await get(`/api/libraries/${libraryId}/authors`)
    return response.data.authors || []
  } catch (error) {
    console.error(`Error fetching authors for library ${libraryId}:`, error)
    return []
  }
}

async function deleteAuthor(authorId: string) {
  const { apiClient } = useApi()
  try {
    await apiClient.value.delete(`/api/authors/${authorId}`)
    return true
  } catch (error) {
    console.error(`Error deleting author ${authorId}:`, error)
    return false
  }
}

export async function executeRemoveEmptyAuthors(formData: Record<string, any>): Promise<ToolResult> {
  try {
    const {
      libraryIds = [],
      deleteWithoutConfirmation = false
    } = formData

    const logs: string[] = []

    const processableLibraries = await fetchLibraryIds(libraryIds);

    logs.push(`Processing ${processableLibraries.length} libraries`)

    let totalAuthorsToDelete = 0
    let totalAuthorsDeleted = 0
    let totalAuthorsDeletionFailed = 0

    for (const libraryId of processableLibraries) {
      logs.push(`\nProcessing library ${libraryId}`)
      
      const authors = await getLibraryAuthors(libraryId)
      const authorsToDelete: Array<{id: string, name: string, numBooks: number}> = []

      // Find authors with 0 books
      for (const author of authors) {
        if (author.numBooks === 0) {
          authorsToDelete.push({
            id: author.id,
            name: author.name,
            numBooks: author.numBooks
          })
        }
      }

      logs.push(`Found ${authorsToDelete.length} authors without books`)
      
      if (authorsToDelete.length === 0) {
        logs.push('No orphaned authors found in this library')
        continue
      }

      logs.push('The following authors will be deleted:')
      for (const author of authorsToDelete) {
        logs.push(`- ${author.name} (${author.numBooks} books) - ID: ${author.id}`)
      }

      totalAuthorsToDelete += authorsToDelete.length
      
      for (const author of authorsToDelete) {
        const success = await deleteAuthor(author.id)
        if (success) {
          logs.push(`Deleted author: ${author.name}`)
          totalAuthorsDeleted++
        } else {
          logs.push(`Error deleting author: ${author.name}`)
          totalAuthorsDeletionFailed++
        }
      }

      logs.push('--- Deletion Complete ---\n')
    }

    logs.push('\n=== SUMMARY ===')
    logs.push(`Total authors found to delete: ${totalAuthorsToDelete}`)
    logs.push(`Total authors successfully deleted: ${totalAuthorsDeleted}`)
    logs.push(`Total authors failed to delete: ${totalAuthorsDeletionFailed}`)

    const message = deleteWithoutConfirmation 
      ? `Successfully processed ${processableLibraries.length} libraries. Deleted ${totalAuthorsDeleted} orphaned authors.`
      : `Found ${totalAuthorsToDelete} orphaned authors across ${processableLibraries.length} libraries. Enable 'Delete Without Confirmation' to actually delete them.`

    return {
      success: true,
      message: message,
      data: logs,
      timestamp: new Date().toISOString()
    }

  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to delete orphaned authors',
      error: error.message || 'Unknown error',
      timestamp: new Date().toISOString()
    }
  }
}